### PR TITLE
feat(widgets): use data param if exist in intWidget()

### PR DESCRIPTION
**Pull request by WANZARGEN**
### To Reviewers
- [ ] Skip (`style`, `chore` ONLY)
- [ ] Not that difficult

### Type of Change
- [ ] New feature
- [ ] Bug fixes
- [x] Feature improvement
- [ ] Refactor
- [ ] Others (performance improvement, refactoring, CI/CD, etc.)

### Checklist
- [x] `Error / Warning / Lint / Type`

### Description

- Upgrade vue version
- Update initWidget() type
- Add WidgetExpose type for defineExpose generic type
- Add data types to all widgets

#### Important Features
- Add widgetDataMap state to dashboard-detail-info store 
- Update DashboardWidgetContainer to use dashboard-detail-info store
- Update DashboardWidgetContainer to give data parameter when calling initWidget() of each widget from widgetDataMap store state
- Add reusePreviousData prop to DashboardWidgetContainer to decide whether to reuse previous data if exists.

### CODE REVIEW

File: src/services/dashboards/dashboard-detail/modules/DashboardWidgetContainer.vue [hunk 1]
```diff
@@ -88,8 +92,8 @@ export default defineComponent<Props>({
                 });
                 return widgetThemeAssigner(widgetThemeOptions);
             }),
-            widgetRef: [] as Array<InstanceType<typeof AWSCloudFrontCost>>,
-            initiatedWidgetMap: {} as {[widgetKey: string]: boolean},
+            widgetRef: [] as Array<WidgetComponent|null>,
```
Comment (sulmoJ) [inline src/services/dashboards/dashboard-detail/modules/DashboardWidgetContainer.vue:100]: I wanted this type...!

File: src/services/dashboards/widgets/type.ts [hunk 2]
```diff
@@ -16,6 +18,6 @@ export interface XYChartData {
 
 export interface Legend {
     name: string;
-    label?: string;
+    label?: TranslateResult;
```
Comment (yuda110) [inline src/services/dashboards/widgets/type.ts:21]: todo: `label` of `Legend` is not `TranslateResult`. It's reference data. (ex. Belkin Snow)

File: src/services/dashboards/dashboard-detail/store/dashboard-detail-info.ts [hunk 3]
```diff
@@ -26,7 +26,9 @@ import type { DashboardModel } from '@/services/dashboards/model';
 import { WIDGET_SIZE } from '@/services/dashboards/widgets/config';
 import { getWidgetConfig } from '@/services/dashboards/widgets/widget-helper';
 
-
+interface WidgetDataMap {
+[widgetKey: string]: any;
```
Comment (yuda110) [inline src/services/dashboards/dashboard-detail/store/dashboard-detail-info.ts:30]: quibble: indent! :)

File: src/services/dashboards/dashboard-detail/modules/DashboardWidgetContainer.vue [hunk 4]
```diff
@@ -141,22 +154,34 @@ export default defineComponent<Props>({
             state.initiatedWidgetMap = initiatedWidgetMap;
         }, { immediate: true, deep: true });
 
-        // for PDF export - start
+
         const refreshAllWidget = async () => {
-            const promises: (()=>void)[] = [];
-            state.widgetRef.forEach((d:any) => {
-                if (typeof d?.refreshWidget === 'function' && state.initiatedWidgetMap[d.$el.id]) promises.push(d?.refreshWidget());
+            const promises: WidgetExpose['refreshWidget'][] = [];
```
Comment (yuda110) [inline src/services/dashboards/dashboard-detail/modules/DashboardWidgetContainer.vue:159]: suggestion: How about `refreshWidgetPromises`?

**Review by yuda110 — APPROVED** [review]
Thanks for your huge refactoring.
I checked widgets quickly, and looked carefully others.

**Review by sulmoJ — APPROVED** [review]
Thank you for your hard work!

**Pull request merged**